### ISSUE TITLE
nn.Sequential should have an add_module(module) instead of add_module(name, module)

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#358]
It should also have a remove_module(index)

**Comment by nhynes** [comment]
How about:
* adding `Module.remove_module(name)`
* adding `Sequential.insert_module(index=None)` (where `None` represents "append")
* overriding `Sequential.remove_module(index=-1)`
* enforcing the invariant that `Sequential._modules` contains contiguous, sequential indices

`Parallel` and `Concat` can then subclass Sequential and haven very simple `forward`s.

If this sounds good, I'll post the PRs. In any case, having the ability to fine tune models would be nice!

**Comment by soumith** [comment]
@nhynes i like your proposed changes. (`Parallel` and `Concat` excluded).
Do post PRs, thank you.

**Comment by apaszke** [comment]
I actually think that considering it immutable is much better. If you want to modify the sequential, just build a new one out of its modules. No need to extend the API for that.

**Comment by soumith** [comment]
no, i dont think it's better to have it be immutable. it's a common operation that one wants to do and expects on a Sequential. let's have a remove_module. 

(I don't see the benefits of immutability here

**Comment by apaszke** [comment]
The benefit is that we're going to end up having `add_module` and `insert_module` that seem to do the same thing, while they're really not. It's going to be confusing.

**Comment by soumith** [comment]
well, i dont mind add_module having a different api for Sequential, but sam said no.

**Comment by apaszke** [comment]
`add_module` makes no sense in context of sequential. It's easy to break its invariants with it. That's why I'm saying that sequentials should be immutable, at least for now. We should redesign this part somehow.

**Comment by nhynes** [comment]
Behold, opinions:

> immutable is much better

I agree with the sentiment and, indeed, the implementation is effectively so. Unfortunately, the rest of the API is mutable and I think that users would be confused by the inconsistency.

> `add_module` makes no sense in context of sequential

Yes; having a separate API would be nice. I think that the real problem is that all modules must have names. In the context of `Parallel` and `Concat`, modules added by the list constructor would also not have meaningful names!

Regarding the PR, I made `insert_module` take both an optional index *and* name since the `OrderedDict` constructor precludes a bijection between names and indices.

**Comment by soumith** [comment]
we'll decide on this this week.
I like @nhynes 's proposed API Change, where name becomes an optional keyword argument. 
We'll have to properly depreciate the old API, as it's used quite often.

**Comment by soumith** [comment]
after debating probably for like more than a month, going back and forth a hundred times, @colesbury vetoed this. So bringing closure to the PR and issue.

His view is that if you want to modify the Sequential, it should be converted to a list, the list be filtered and the Sequential re-constructed, instead of pretending to semi-mutate the module with some methods. And the API change to `add_module` was voted out (i.e. name becoming a kwarg) for many reasons, backward-compat being one.

**Comment by nhynes** [comment]
Thanks for getting back to this :) I was worried that it got tabled but didn't want to distract from more pressing issues!

**Comment by nicholas-leonard** [comment]
@soumith I disagree with this decision.

Coming from Lua Torch, I was expecting this to work:
```
model = nn.Sequential(nn.Embedding(opt.vocabsize, opt.inputsize))

inputsize = opt.inputsize
for rnnsize in opt.rnnsize:
	model.add_module(nn.LSTM(inputsize, rnnsize, 1))
	inputsize = rnnsize

for linearsize in opt.linearsize:
	model.add_module(nn.Linear(inputsize, linearsize))
	model.add_module(nn.ReLU())
	inputsize = rnnsize
```
Instead, I had to write this
```
layers = [nn.Embedding(opt.vocabsize, opt.inputsize)]

inputsize = opt.inputsize
for rnnsize in opt.rnnsize:
	layers.append(nn.LSTM(inputsize, rnnsize, 1))
	inputsize = rnnsize

for i, linearsize in enumerate(opt.linearsize):
	layers.append(nn.Linear(inputsize, linearsize))
	layers.append(nn.Sigmoid() if i == (len(opt.linearsize)-1) else nn.ReLU())
	inputsize = linearsize

model = nn.Sequential(*layers)
```
I hope their is a good reason for this design change.

**Comment by lianshushu** [comment]
I happened a situation . How could I do if I want to add or remove or  change one conv layer or just change the stride of it in a pretrained resnet model which I still need the pretrained parameters. Thank you so much.

**Comment by lianshushu** [comment]
How do I reuse the pretrained weight if I rebuild a sequential

**Comment by nhynes** [comment]
If you want to just change the stride/padding/dilation, you can just directly change the [property](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/conv.py#L22) on the `ConvNd` object.

If you want to reuse the pretrained weight, just do `new_layer.weight = orig_layer.weight` and then use the `new_layer` when constructing the new `Sequential`.

**Comment by lianshushu** [comment]
Thank you . I see.

**Comment by pbloem** [comment]
Sorry to rake this one up again, but having to name each module in a Sequential model when I add them one-by-one really ruins the readability of my code.

Would it be a solution to add an extend() to the sequential model (with the same contract as that in ModuleList)?

This would allow you to write
```
seq.extend([Conv2d(3, 5, 4)])
```
for a single module and has the benefit of expanding to 
```
seq.extend([
    Conv2d(3, 5, 4),
    RelU(),
    MaxPool2d((4, 4)),
])
```
for multiple modules. Which looks a lot like how you would add a list of modules to the constructor. 

This doesn't affect add_module, so it doesn't hurt any of the existing contracts and backward compatibilities, and it stops you from having to give arbitrary, unique names to your modules.

**Comment by fmassa** [comment]
For you use-cases, can't you put the modules in a list and just use the constructor of `nn.Sequential`?

**Comment by pbloem** [comment]
Fair enough, that would work too. It's a little less direct, but not a big deal. With the current code that's certainly the most readable option.

**Comment by apaszke** [comment]
@fmassa that's true, but what if the `Sequential` isn't coming from you? I think `list(seq) + my_modules` might work, but I'm not 100% sure, and not everyone would come up with that. I wouldn't be opposed to having `.extend`.

**Comment by songyuc** [comment]
@nhynes , hi, Nick,
I see that the nn.Sequential.add_module() still require a `name` parameter.
So how is the proposal like `add_module(module, name=None)`?
